### ISSUE TITLE
krun -- pretty printer: an empty list

### ISSUE THREAD

**Issue by daejunpark** [kframework/k-legacy#188]
The pretty printer of krun prints an unnecessary empty list.

For example, for the given K definition:

```
  syntax Pgm ::= "{" Exps "}"
  syntax Exps ::= List{Int,","}
```

and a program:

```
{ }
```

krun prints as follows:

```
{ .List{","} }
```

How can I remove the annotation? Or, how can I make K parse the annotated program?

Indeed, my scenario is that there exist two K definitions, where
- the first K definition translates a given program to a simpler form,
- and, the second K definition reads the translated program and does something.

In that case, I need a way to either remove the annotation (because K parser cannot parse that annotation), or make the K parser parse that annotation.

Thanks in advance!

**Comment by liyili2** [comment]
I am not sure if this is a krun problem. I know this problem before, and I
think it is because of the normalizing procedure in the k rewriter, but
what is the expected result you want?

If you want the result to be { Exps } something, I think it is the matter
of parser.

If you want the result to be { }, so what is the type of this program? what
is the difference between { } and { .List }, since .List is considered to
be null in K

Liyi Li

**Comment by grosu** [comment]
Liyi,

You should really talk to Eric about this.  He has already gone through all this.  In short, we want to pretty-print no .List{Int,","} terms anymore.  Instead, we should either print .Exps or whatever is the actual concrete unit that it represents when it appears alone, or to simply drop it when it appears as part of a non-empty list.  It mostly works, but there are some cases where it does not.

Grigore

**Comment by liyili2** [comment]
Yes, I have already sent an email to eric and asked for help with a
meeting. I am not sure if he is in the USA or in other country, but I will
talk with him anyway next week.

Liyi Li

**Comment by dwightguth** [comment]
Ideally what we want is to tell krun to unparse a term in a particular module using only syntax from that module, and to report an error if this is not possible. This is something that should eventually be handled as part of the new KORE implementation of the unparser.
